### ISSUE TITLE
runtime: use the STM32 M4 MPU protection regions for stack guards

### ISSUE THREAD

**Issue by eliasnaur** [tinygo-org/tinygo#4057]
I just spent some time debugging a program crash that was caused by a stack overflow. Tinygo uses a stack canary, but only check it during task switch. Can the M4 MPU regions be used to protect the canary, thus triggering a hardware fault immediately on overwriting it?

I started writing the assembly according to [the manual](https://www.st.com/resource/en/programming_manual/pm0214-stm32-cortexm4-mcus-and-mpus-programming-manual-stmicroelectronics.pdf):

```diff
diff --git a/src/internal/task/task_stack_cortexm.S b/src/internal/task/task_stack_cortexm.S
index fe1f44e0..bc63174d 100644
--- a/src/internal/task/task_stack_cortexm.S
+++ b/src/internal/task/task_stack_cortexm.S
@@ -102,10 +102,27 @@ tinygo_swapTask:

     // Switch the stack. This could either switch from PSP to MSP, or from MSP
     // to PSP. By using an XOR (eor), it will just switch to the other stack.
+    // Load address of stack canary.
+    add  r6, r3, #0x8
     mrs  r0, CONTROL // load CONTROL register
     movs r3, #2
     eors r0, r0, r3  // flip the SPSEL (active stack pointer) bit
     msr  CONTROL, r0 // store CONTROL register
+    // Set up protected region for the stack canary.
+    ldr  r4, #0xE000ED98 // MPU_RNR
+    // Set region #0 and disable it.
+    ldr  r5, #0
+    str  r4, [r5, 0x0]
+    str  r5, [r4, 0x8]
+    // Set base address.
+    str  r6, [r4, 0x4]
+    // Set read-only attribute.
+    ldr  r5, #0x7
+    str  r5, [r4, 0xa]
+    // Set size and enable.
+    ldr  r5, #0x5
+    str  r5, [r4, 0x8]
+    dsb              // Required to flush the MPU settings.
     isb              // required to flush the pipeline

     // Load state from new task and branch to the previous position in the
```

but haven't figured out how to plumb through the canary address to `swapTask`.

Is this a reasonable feature? If so, where best to put it? Should the canary guard address be passed to `swapTask`, or can it be computed from the stack pointer alone?

**Comment by aykevl** [comment]
Yes, this would be great to have! If you're still up for implementing this :)
Also would be great to print an error message when the stack overflows.

How small can these protected regions be? (Of course we'd like to keep it very small to not waste too much memory).

I think adding a third parameter for the end of stack world be reasonable, though I'm not entirely sure whether that would be easy to do with all the different chips and architectures we have.

**Comment by eliasnaur** [comment]
I won't have time to implement this anytime soon.

Ideally the outcome of such as feature should not be a crash, but rather a stack resize, similar to how Big Go does it. TinyGo stacks are both too small (leading to mysterious crashes in stack hungry goroutines) and too large (a large `-stacksize` wastes memory for every goroutine for which TinyGo can't deduce a tight stack bound)

**Comment by aykevl** [comment]
Stack resizes would be nice, but are not currently possible with LLVM (since LLVM doesn't keep track of which values on the stack or in registers are pointers to the stack and therefore need to be updated when the stack is moved).
